### ISSUE TITLE
Neighbor Patch Filtering and Target-Based Indexing Lead to Mismatched Spot Counts and Index Errors

### ISSUE THREAD

**Issue by NozomiMizore** [NEXGEM/TRIPLEX#15]
## Problem Source
- Neighbor patches are generated with stricter QC (larger patch size, WSI bounds, tissue mask), so some spots that exist in target are dropped only from neighbor.
- The matching step `match_to_target` aligns neighbor to target, but the rest of the pipeline (especially .h5ad and training indices) is still driven by target, not by neighbor.
- During training, indices are derived from .h5ad (== target), then used to index neighbor embeddings; on slides where len(neighbor) < len(adata/target) , such as **CCRCC INT1** and **INT4**，this causes out-of-bounds errors.

## Relevant Code
1. Neighbor patches use a larger patch size and separate filtering
```python
# src/preprocess/prepare_data.py

def save_patches(name, input_dir, output_dir,
                 platform='visium', save_targets=True,
                 save_neighbors=False, num_n=5):

    ...

    if save_neighbors:
        if os.path.exists(f"{output_dir}/patches/neighbor/{name}.h5"):
            print("Neighbor patches already exists. Skipping...")
        else:
            if st._tissue_contours is None:
                print("Segmenting tissue...")
                st.segment_tissue(method='deep')

            print("Dumping neighbor patches...")
            st.dump_patches(
                f"{output_dir}/patches/neighbor",
                name=name,
                target_patch_size=224*num_n,  # much larger neighbor patches
                target_pixel_size=dst_pixel_size,
                threshold=int(0.15 // num_n),
                dump_visualization=False
            )

            print("Matching neighbor patches to target patches...")
            target_path = f"{output_dir}/patches/{name}.h5"
            neighbor_path = f"{output_dir}/patches/neighbor/{name}.h5"

            match_to_target(target_path, neighbor_path)
```
2. Matching step: neighbor is trimmed to the target intersection (target stays unchanged)
```python
# src/preprocess/prepare_data.py

def save_patches(name, input_dir, output_dir,
                 platform='visium', save_targets=True,
                 save_neighbors=False, num_n=5):

    ...

    if save_neighbors:
        if os.path.exists(f"{output_dir}/patches/neighbor/{name}.h5"):
            print("Neighbor patches already exists. Skipping...")
        else:
            if st._tissue_contours is None:
                print("Segmenting tissue...")
                st.segment_tissue(method='deep')

            print("Dumping neighbor patches...")
            st.dump_patches(
                f"{output_dir}/patches/neighbor",
                name=name,
                target_patch_size=224*num_n,  # much larger neighbor patches
                target_pixel_size=dst_pixel_size,
                threshold=int(0.15 // num_n),
                dump_visualization=False
            )

            print("Matching neighbor patches to target patches...")
            target_path = f"{output_dir}/patches/{name}.h5"
            neighbor_path = f"{output_dir}/patches/neighbor/{name}.h5"

            match_to_target(target_path, neighbor_path)
```
Effect: neighbor → [target ∩ neighbor], but target remain larger. 
3. ST (adata) is aligned to target patches, not to neighbor
```python
# src/preprocess/prepare_data.py

def preprocess_st(name, adata, output_dir):
    os.makedirs(f"{output_dir}/adata", exist_ok=True)
    save_dir = f"{output_dir}/adata/{name}.h5ad"

    if os.path.exists(save_dir):
        print(f"ST data already exists for {name}. Skipping...")
        return None
    print(f"Saving ST data for {name}...")

    print("Dumping matched ST data...")
    with h5py.File(f"{output_dir}/patches/{name}.h5", "r") as f:
        barcode = f['barcode'][:].astype('str').squeeze()

    barcode = pd.DataFrame(index=barcode)
    barcode_merged = pd.merge(
        adata.obs, barcode, left_index=True, right_index=True
    ).index
    adata = adata[barcode_merged]

    adata.write(save_dir)
    return adata
```
Resulting relation: [adata == target], but neighbor is a strict subset of target on some slides.
4. Training indices are driven by h5ad (== target), then reused for neighbor
```python 
# src/dataset/base_dataset.py (indexing based on h5ad length)

class STDataset(torch.utils.data.Dataset):
    def __init__(...):
        ...
        if phase == 'train':
            self.adata_dict = {
                _id: self.load_st(_id, self.genes, **self.norm_param)
                for _id in self.ids
            }
            self.lengths = [len(adata) for adata in self.adata_dict.values()]
            self.cumlen = np.cumsum(self.lengths)

    def __getitem__(self, index):
        if self.phase == 'train':
            i = 0
            while index >= self.cumlen[i]:
                i += 1
            idx = index
            if i > 0:
                idx = index - self.cumlen[i-1]

            name = self.int2id[i]
            img = self.load_img(name, idx)
            adata = self.adata_dict[name]
            expression = adata[idx].X
            ...
```
```python
# src/dataset/tri_dataset.py (same idx used for neighbor embeddings)

class TriDataset(STDataset):
    def __getitem__(self, index):
        if self.phase == 'train':
            ...
            name = self.int2id[i]
            img = self.load_img(name, idx)
            img = self.transforms(img)

            # idx comes from h5ad / target length
            neighbor_emb, mask = self.load_emb(
                name, emb_name='neighbor', idx=idx
            )
            adata = self.adata_dict[name]
            expression = adata[idx].X
            ...
```
On samples where len(neighbor) < len(adata/target), idx is valid for adata but out-of-bounds for neighbor, causing runtime index errors.
## Solution
Additionally, use neighbor barcode as the reference set, and intersect it with target patches, global embeddings, and .h5ad barcodes. So after this alignment, for every sample: `len(adata) == len(target patches) == len(neighbor patches) == len(global embeddings)`
